distributed/safetensor_utils.py

```diff
@@ -0,0 +1,325 @@
+        weight_map = {k.removeprefix("model."): v for k, v in weight_map.items()}
+    return state_dict, weight_map
+
+
+def get_needed_files(
+    state_dict: Dict[str, torch.Tensor], weight_map: Dict[str, str]
+) -> Set[str]:
+    needed_files = set()
+    for param in state_dict.keys():
+        file = weight_map.get(param)
+        if file:
+            needed_files.add(file)
+        elif param.endswith("weight"):
+            logger.warning(
+                f"Parameter {param} not found in weight map, please check..."
+            )
+            raise ValueError(f"Missing file for {param} in {weight_map.keys()}")
+    logger.info(f"Needed files: {needed_files}")
+    return needed_files
+
+
+def load_checkpoint(full_path: str, device: torch.device) -> Dict[str, torch.Tensor]:
+    tensors = {}
+    with safe_open(full_path, framework="pt", device=device) as f:
+        for k in f.keys():
+            tensors[k] = f.get_tensor(k)
+    logger.info(f"Loaded {len(tensors)} tensors from {full_path}")
+    return tensors
+
+
+def update_state_dict(
+    state_dict: Dict[str, torch.Tensor],
+    checkpoint: Dict[str, torch.Tensor],
+    weight_map: Dict[str, str],
+    new_to_old_keymap: Dict[str, str],
+    file: str,
+    updated_states: Set[str],
+    device: torch.device,
+):  
+    count_dtensors_loaded = 0
+    for param, file_with_param in weight_map.items():
+        if file_with_param == file and param in state_dict:
+            model_param = (
+                "output.weight" if param == "output.weight" else f"model.{param}"
+            )
+            old_param = new_to_old_keymap.get(model_param)
+
+            if old_param not in checkpoint:
+                logger.warning(f"Missing {old_param} in checkpoint")
+                continue
+
+            checkpoint_tensor = checkpoint[old_param]
+            stage_tensor = state_dict[param]
+            
+            # stage_is_dtensor = is_dtensor(stage_tensor)
+            # logger.info(f"cme DType Check: {param=}, {stage_is_dtensor=}, {checkpoint_tensor.dtype=}, {stage_tensor.dtype=}")
+            
+            checkpoint_tensor = compare_and_reverse(stage_tensor, checkpoint_tensor)
+            
+            # here we need to check if the tensor is a DTensor and if so, adjust the 
+            # shape and placement to match the model DTensor.  
+            if is_dtensor(stage_tensor):
+                model_tensor = load_into_dtensor(checkpoint_tensor, stage_tensor)
+                # logger.info(f"DTensor: Loaded {param} into {model_tensor=}")
+                
+                state_dict[param] = model_tensor
+                count_dtensors_loaded += 1
+                
+            else:
+                # regular tensor, just update directly
+                checkpoint_tensor = checkpoint_tensor.to(device)
+                state_dict[param] = checkpoint_tensor
+
+            state_dict[param] = state_dict[param].to(checkpoint_tensor.dtype)
```

Comment:
wondering the purpose of this to() call. 
It seems we just updated `state_dict` with `checkpoint_tensor` above.

Reply:
ah, this is b/c I have two branches, one for dtensors and one for regular tensors...let me just consolidate it as we only need this last one and not the one special cased under regular tensors.